### ISSUE TITLE
[contracts] Gas estimation for some cross-contract calls too low

### ISSUE THREAD

**Issue by cmichi** [paritytech/substrate#9334]
It's a bit tricky to reproduce, I tried reducing it to a simpler example, but haven't managed to.

You have to use the [`ink-waterfall`](https://github.com/paritytech/ink-waterfall) to reproduce. For this you have to install [the dependencies of that project](https://github.com/paritytech/ink-waterfall#required-dependencies). This is most likely only `geckodriver` and possibly the latest `canvas-node` for you.

After that start `canvas-node` in one terminal:
```
canvas --dev --tmp -lerror,runtime::contracts=debug
```

And in another the test:
```
git clone --branch cmichi-gas-estimation-bug-with-invoked-payable-transaction https://github.com/paritytech/ink-waterfall.git
cd ink-waterfall/

# Adapt the path to your system
export INK_EXAMPLES_PATH=~/projects/ink/examples/ 

# Firefox will stay open after the test run, enabling you to do further debugging
export WATERFALL_CLOSE_BROWSER=false

RUST_LOG=info cargo test --no-default-features --features polkadot-js-ui -- --test-threads=1 multisig_works_with_payable_transaction
```

If the bug can be reproduced correctly you will see an `OutOfGas` error at the end of the test. On the console this will show up as
```
[multisig_works_with_payable_transaction invokeTransaction] found status message "contracts.OutOfGas"
```

What the test does can be viewed [here](https://github.com/paritytech/ink-waterfall/blob/cmichi-gas-estimation-bug-with-invoked-payable-transaction/src/tests/multisig_plain.rs#L105-L106). This is a special branch which I created for this bug report, the difference to `master` is that it additionally sets `max_gas` explicitly to work around the `OutOfGas`. Also I commented out the assertion for log messages at the end of the test.

The tl;dr is that the test
* …deploys ink!'s [multisig_plain](https://github.com/paritytech/ink/tree/master/examples/multisig_plain) example
* …submits a transaction to it
* …the transaction is a call to ink!'s [contract-transfer::was_it_ten](https://github.com/paritytech/ink/blob/master/examples/contract-transfer/lib.rs#L86-L87)
* …the transaction is confirmed by different parties and then invoked

__Edit:__ I just found this: If I remove the `ink_env::debug_println!` from [contract-transfer::was_it_ten](https://github.com/paritytech/ink/blob/master/examples/contract-transfer/lib.rs#L88-L91) there is no longer an `OutOfGas` error! So it seems that it might not even be related to the `payable`.

**Comment by athei** [comment]
The issue is indeed caused by `debug_println!` as you already pointed out in your edit. More specifically by [this tiny branch](https://github.com/paritytech/ink/blob/a2e549eaee5175ad98aa79cc569628d34496f4d0/crates/env/src/engine/on_chain/ext.rs#L674) which is not ran during RPC execution but then later run when executed as transaction making the estimation too low. So the issue arises from ink! executing more code when executes as transaction. When you have more than one `debug_println!` the issue does not occur because the missing gas gets recouped by not calling into the supervisor after the first time.

On a side note: The reproducer here is bigger than necessary. One can simply reproduce it by deploying the transfer contract on canvas UI and call `was_it_ten` directly while supplying 10 pico. No need to go through multisig. This won't work on polkadot-js because that UI **never** uses gas estimation for non mutating (i.e getter) messages. It just uses a default value in this case. This is probably why you only noticed this issue when `was_it_ten` is called through some other contract.

Another issue that plagues both UIs and that makes this harder to reproduce it that they don't use the gas estimation whenever the execution finishes unsuccessfully. Which is the case when something other is supplied than 10 pico. It is hard to detect because both UIs don't clear the estimation field or throw an error but silently "ungrey" the gas estimation text field leaving whatever value in there which happens to be there before (often put there from another message). Please note that there is no apparent reason why the UIs behave this way. Gas estimation is returned by the RPC unconditionally (not depending on the success of the execution).
